File: c++/c14_1_graph_traversal.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	void graph_traversal(const vector<vector<ll>>& adj, ll x, auto f) {
		auto dfs = [&](this auto self, ll s, ll e) -> void {
			// process node s
			f(s);
			for (auto u : adj[s]) {
				if (u != e) self(u, s);
			}
			};
		dfs(x, 0);
	}
	vector<ll> dynamic_programming(const vector<vector<ll>>& adj, ll x) {
		vector<ll> count(adj.size(), 1);
		auto dfs = [&](this auto self, ll s, ll e) -> void {
			for (auto u : adj[s]) {
				if (u == e) continue;
				self(u, s);
				count[s] += count[u];
			}
			};
		dfs(x, 0);
		return count;
	}
}

TEST(C141GraphTraversal, graphTraversal) {
	vector<ll> ret;
	graph_traversal({ {},{} }, 1, [&](ll s) { ret.push_back(s); });
	EXPECT_EQ(ret, vector<ll>{1});
	ret.clear();
	graph_traversal({ {},{2,3,4},{1,5,6},{1},{1,7},{2},{2,8},{4},{6} }, 1, [&](ll s) { ret.push_back(s); });
	EXPECT_EQ(ret, (vector<ll>{1,2,5,6,8,3,4,7}));
}

TEST(C141GraphTraversal, dynamicProgramming) {
	EXPECT_EQ(dynamic_programming({ {},{} }, 1), (vector<ll>{ 1,1 }));
	EXPECT_EQ(dynamic_programming({ {},{2,3,4},{1,5,6},{1},{1,7},{2},{2,8},{4},{6} }, 1), (vector<ll>{ 1,8,4,1,2,1,2,1,1 }));
}

File: python/c14_1_graph_traversal.py
import unittest
import sys

sys.setrecursionlimit(max(sys.getrecursionlimit(),1<<20))

def graph_traversal(adj,x):
    def dfs(s,e):
        yield s
        for u in adj[s]:
            if u!=e:
                yield from dfs(u,s)
    return dfs(x,0)

def dynamic_programming(adj,x):
    count=[1]*len(adj)
    def dfs(s,e):
        for u in adj[s]:
            if u==e:
                continue
            dfs(u,s)
            count[s]+=count[u]
    dfs(x,0)
    return count

class Test(unittest.TestCase):
    def test(self):
        adj=[[],[2,3,4],[1,5,6],[1],[1,7],[2],[2,8],[4],[6]]
        self.assertEqual([*graph_traversal([[],[]],1)],[1])
        self.assertEqual([*graph_traversal(adj,1)],[1,2,5,6,8,3,4,7])
        self.assertEqual(dynamic_programming([[],[]],1),[1,1])
        self.assertEqual(dynamic_programming(adj,1),[1,8,4,1,2,1,2,1,1])

if __name__=='__main__':
    unittest.main()
